Optimise lock behaviour for GuessVerificationProgress()

jonasschnelli opened the pull request:
`GuessVerificationProgress()` needs `cs_main` due to accessing the `pindex->nChainTx`.
This adds a `AssertLockHeld` in `GuessVerificationProgress()` and adds the missing locks in...
* `LoadChainTip()`
* `ScanForWalletTransactions()` (got missed in #11281)
* GUI, `ClientModel::getVerificationProgress()` <--- **this may have GUI performance impacts**, but could be relaxed later with a cache or something more efficient.

jonasschnelli committed:
Fix missing cs_main lock for GuessVerificationProgress()

jonasschnelli commented:
Not meant for 0.16.

practicalswift commented:
Related:
* #11754 ("wallet: Add missing cs_wallet locks when accessing m_last_block_processed")
* #11694 ("rpc: Add missing cs_main lock in getblocktemplate(...)")
* #11652 ("Add missing locks to init.cpp (in AppInitMain + ThreadImport) and validation.cpp")

promag commented on src/qt/clientmodel.cpp:
```diff
@@ -143,6 +143,7 @@ double ClientModel::getVerificationProgress(const CBlockIndex *tipIn) const
         LOCK(cs_main);
         tip = chainActive.Tip();
     }
+    LOCK(cs_main);
```
Just move the above lock up?

promag reviewed (commented):
utACK.

ryanofsky commented on src/validation.cpp:
```diff
@@ -4659,7 +4662,9 @@ bool DumpMempool(void)
 //! Guess how far we are in the verification process at the given block index
+//! needs cs_main due to access of pindex->nChainTx (mutable)
 double GuessVerificationProgress(const ChainTxData& data, const CBlockIndex *pindex) {
+    AssertLockHeld(cs_main);
```
I think maybe the new lock and assert here could be dropped and comment above could be changed to: "require cs_main if pindex has not been validated yet (because nChainTx might be unset)."

If we're calling this on blocks from chainActive, it seems like locking should be unnecessary.

promag replied:
After 3d40869, `ClientModel::getVerificationProgress` is the only caller without the lock, so `AssertLockHeld` seems fine to me.

promag commented on src/validation.cpp:
```diff
@@ -3803,10 +3803,13 @@ bool LoadChainTip(const CChainParams& chainparams)
-        DateTimeStrFormat("%Y-%m-%d %H:%M:%S", chainActive.Tip()->GetBlockTime()),
-        GuessVerificationProgress(chainparams.TxData(), chainActive.Tip()));
+    {
+        LOCK(cs_main);
```
The whole function should lock `cs_main`. However this is called only called in `AppInitMain`, and #11041 adds the lock [there](https://github.com/bitcoin/bitcoin/pull/11041/files#diff-c865a8939105e6350a50af02766291b7).

If this goes first, I suggest cherry pick 3d40869 instead.

promag commented:
@jonasschnelli please see 6e36821 which came after discussing with @ryanofsky on IRC.

jonasschnelli commented:
@promag https://github.com/bitcoin/bitcoin/commit/6e36821823b498c307787543ad59b9e9ccffcc63 seems good work. I'll propose to do this in a separate PR.

jonasschnelli commented:
Changed my mind. Included https://github.com/bitcoin/bitcoin/commit/6e36821823b498c307787543ad59b9e9ccffcc63.
* Fixed @ryanofsky point about `AssertLockHeld()`
* Followed @promag's advice and moved cs_main lock up (https://github.com/bitcoin/bitcoin/pull/12287#discussion_r164444275)

promag commented:
Thread https://github.com/bitcoin/bitcoin/pull/12287#discussion_r164570542

In this case, where `nChainTx` may require the lock and `GuessVerificationProgress()` is always called with the lock, I'd prefer to assert the lock and be *correct* like @gmaxwell said in IRC.

Other than that, utACK 90ba2df.

ryanofsky reviewed (commented):
Nice code simplifications and comments.

Conditional utACK 90ba2df11b5bc943ac48b49b5da8023864dc842d if PR title ("Fix missing cs_main lock for GuessVerificationProgress") is updated, because the added locking is just defensive, not because locks are missing. Also think adding the assert would be misleading for the same reason.

jonasschnelli commented:
Changed the PR title.

promag commented:
Tested ACK 90ba2df.

Let's get this merged? 🙄
